Use account identities as default petnames

dbrans opened the pull request:
## **Description**

Listens for changes to account identities and sets those as petnames.

## **Related issues**

Part of: https://github.com/MetaMask/MetaMask-planning/issues/1624 (https://github.com/orgs/MetaMask/projects/47/views/1?pane=issue&itemId=43756188)

Follow-up: migration to grab any existing account identities.

## **Manual testing steps**

1. Open MM extension
2. Change the account label for any account
3. TODO

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

## **Pre-merge author checklist**
- [x] how to confirm in the UI that petname was added?
- [x] What is the correct type and variation to use for the setName call?
- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dbrans committed:
snapshot

dbrans commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,90 @@
+};
+
+const ACCOUNT_LABEL_NAME_TYPE = NameType.ETHEREUM_ADDRESS;
+const ACCOUNT_LABEL_CHAIN_ID = CHAIN_IDS.MAINNET;
```
What is the correct NameType and variation to use for the NameController::setName calls?

dbrans committed:
Added tests

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge remote-tracking branch 'origin/develop' into dbrans/petnames-accountids

dbrans committed:
Added migration

dbrans committed:
Updated migrations/index.js

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,105 @@
+  name: string;
+};
+
+export type PertinentState = {
```
As we're exporting this type, should we give it more context so it's meaningful in other files without an alias?

Such as `PreferencesControllerState`?

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,105 @@
+  };
+};
+
+const MAINNET_CHAIN_ID = '0x1';
```
This highlights an important consideration.

Petnames are stored in the `NameController` with unique names and proposed names per variation. In the case of ethereum address petnames, these variations are the chain IDs. This means we currently need to create the new petname entries for every applicable chain. To achieve this, we can either:

1. Update the listener to still fire on the preferences controller state change but also whenever the network changes, at which point we can create the relevant petnames to match the account identities for the current chain. The downside of this is we grow and duplicate the state alot over time if the user interacts with many networks, plus it adds complexity to this bridge logic.

2. Use a special variation for these petnames such as `*` and update the petnames UI components to check that variation as a fallback if it can't find a petname for a specific chain ID. This means we only have to define these petnames once so the state is smaller, plus no listening to network changes, plus a much simpler migration as we only have to worry about this generic variation (`*`) rather than every relevant network.

In both of these solutions, at least for now, the petnames UI will contain to only override the entry for the current chain, rather than the generic entry. Perhaps in future we could let users choose to save a new petname for all chains or a single one.

dbrans replied:
I've added FALLBACK_VARIATION to the useName hook. PTAL.

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,105 @@
+ * @param oldEntries - The last seen IdentityEntries in the preferences controller.
+ * @param newEntries - The new IdentityEntries.
+ */
+function groupEntries(
```
As this logic is very similar to the `AddressBookPetnamesBridge`, is it worth creating an abstract `PetnamesBridge` which we could re-use for both the address book and account identities to simplify them both and reduce duplication, plus simplify the unit testing.

It could just be the existing `AddressBookPetnamesBridge` logic but with the source of the data abstracted through abstract methods such as `onDataSourceChange`, `getSourceEntries`, and `updateSourceEntries`.

We could also use a constructor argument such as `isBidirectional` to control whether a specific bridge cares about updating the source, as done with the address book but not the account identities.

matthewwalsh0 commented on app/scripts/migrations/106.ts:
```diff
@@ -0,0 +1,73 @@
+
+    const normalizedAddress = address.toLowerCase();
+    const nameEntry = names[normalizedAddress] ?? {};
+    const petnameExists = Boolean(nameEntry[ACCOUNT_LABEL_CHAIN_ID]?.name);
```
We'll likely have to use the default variation here so we create all the relevant petnames only once for existing account identities.

dbrans replied:
Done.

matthewwalsh0 commented on app/scripts/migrations/106.test.ts:
```diff
@@ -0,0 +1,251 @@
+      ...oldState,
+      NameController: {
+        names: {
+          [ACCOUNT_LABEL_NAME_TYPE]: {},
```
Minor, but for the sake of stability and the minimal changes possible, should we not create this `NameController` entry in the state if there have been no entries created?

matthewwalsh0 commented on app/scripts/migrations/106.test.ts:
```diff
@@ -0,0 +1,251 @@
+    expect(transformedState.data).toEqual(oldState);
+  });
+
+  const ADDRESS_1 = '0xc0ffee254729296a45a3885639AC7E10F9d54979';
```
Minor, but should we put all the constants at the top of the file for visibility? I think that's our standard elsewhere.

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.test.ts:
```diff
@@ -0,0 +1,153 @@
+const NAME_MOCK = 'testName';
+const NAME_2_MOCK = 'testName2';
+
+interface MockPreferencesController
```
Minor, but I find the simplest way to type the mocks in the unit tests is to set the return type to `jest.Mocked<SomeType>` and then just set the minimal requited properties and cast to `any`, as you've done with `createNameControllerMock`.

matthewwalsh0 replied:
I see what you mean. I think usually we'd just create a dedicated variable so we're only dealing with mocking functions at one level.

So in the `beforeEach`, we could have:

`preferencesControllerStoreMock = jest.mocked(preferencesControllerDefault.store);`

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,105 @@
+
+const MAINNET_CHAIN_ID = '0x1';
+export const ACCOUNT_LABEL_NAME_TYPE = NameType.ETHEREUM_ADDRESS;
+export const ACCOUNT_LABEL_VARIATION = MAINNET_CHAIN_ID;
```
`NameType.ETHEREUM_ADDRESS` is correct as it's petnames specifically for Ethereum addresses we're concerned with, but the chain ID is a great question that I've discussed in a separate comment.

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
snapshot

dbrans committed:
Abstract out the generic and name bridge code

dbrans committed:
Merge remote-tracking branch 'origin/develop' into dbrans/petnames-accountids

matthewwalsh0 commented on app/scripts/lib/EntryBridge.ts:
```diff
@@ -0,0 +1,201 @@
+ * @param applyChangeToDestination - A function that applies a change to the
+ * destination.
+ */
+export type OneWayBridgeConfig<Entry> = {
```
I'm really impressed with what you've done here by abstracting the concept of a bridge itself, but I fear that at this level of abstraction that there is a trade off in additional complexity and code plus I worry it's unlikely other use cases would re-use this given it's still specific to our bridging use cases.

My intent was just for the concept of a "Petnames Bridge" to be abstracted meaning we avoid duplication between the two sources we now have (address book and account identities) but still encapsulate all our logic in petnames specific files.

matthewwalsh0 commented on app/scripts/lib/EntryBridge.ts:
```diff
@@ -0,0 +1,201 @@
+ */
+export type OneWayBridgeConfig<Entry> = {
+  comparators: EntryComparators<Entry>;
+  getSourceEntries: () => Entry[];
```
Since the only purpose of these bridges is to synchronise data, would it be simpler and more idiomatic to use the template pattern with an abstract class and provide the variations in sources through abstract methods?

Very rough, but perhaps something like:

```
abstract class AbstractPetnamesBridge {
    constructor({isTwoWay}: {isTwoWay: boolean})
    init() // Adds initial listeners
    abstract getSourceEntries(): NameEntry[]
    abstract updateSourceEntries(type: ChangeType, entry: NameEntry)
}

class AddressBookPetnamesBridge extends AbstractPetnamesBridge
class AccountIdentitiesPetnamesBrige extends AbstractPetnamesBridge
```

All the logic would be specific to petnames meaning we can simplify the implementation using concrete types like `NameEntry` and avoiding the need for `EntryComparators` since we can hardcode our `NameEntry` comparisons.

matthewwalsh0 commented on app/scripts/lib/EntryBridge.ts:
```diff
@@ -0,0 +1,201 @@
+/**
+ * The type of change that occurred.
+ */
+export type ChangeType = 'added' | 'updated' | 'deleted';
```
This type is a great idea as it's likely the source updates will be very similar for add and update.

Though would it be cleaner to use an enum to define these supported values rather than a string union?

dbrans replied:
Done in next update.

matthewwalsh0 commented on app/scripts/lib/EntryBridge.ts:
```diff
@@ -0,0 +1,201 @@
+/**
+ * A callback that applies a change to an entry to a destination.
+ */
+export type ApplyChangeCallback<Entry> = (type: ChangeType, entry: Entry) => void;
```
If we go with the abstract class approach, we shouldn't need a type for this as it is enforced by the abstract method signature.

dbrans replied:
We could still could use this type when we apply the changelist to either side: 

Here is #applyChangeList _with_ the callback type:
```typescript
#applyChangeList(
    changeList: ChangeList,
    applyChange: ApplyChangeCallback,
  ): void {
    for (const type of Object.values(ChangeType)) {
      for (const entry of changeList[type]) {
        applyChange(type, entry);
      }
   }
}
```

And without:
```typescript
#applyChangeList(
    changeList: ChangeList,
  ): void {
    for (const type of Object.values(ChangeType)) {
      for (const entry of changeList[type]) {
        if (direction === 'Source->NameController')
          this.#updateNameControllerEntry(type, entry);
        } else {
          this.#updateSourceEntry(type, entry);
        }
      }
   }
}

```

matthewwalsh0 commented on app/scripts/lib/EntryBridge.ts:
```diff
@@ -0,0 +1,201 @@
+};
+
+// Manages a two-way bridge for entries, preventing recursive updates
+export class TwoWayBridge<Entry> {
```
As mentioned in the other comment, I think we would simplify out logic if we define both the one way and two way bridges in a single abstract class and use a constructor boolean argument to request two way support or not.

matthewwalsh0 commented on app/scripts/lib/NameBridge.ts:
```diff
@@ -0,0 +1,144 @@
+
+type AllowedEvents = NameStateChange;
+
+export type BridgeMessenger<Namespace extends string> =
```
This is really nice technically, but for our purposes, we can just encapsulate and hardcode the `NameController` state subscription inside our `AbstractPetnamesBridge` given a specific `messenger` in the constructor argument.

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,46 @@
+ * @param preferencesController - The preferences controller to listen to.
+ * @param nameController - The name controller to update.
+ */
+export default function setupAccountLabelsPetnamesBridge(
```
If we use an abstract class, then this entire file would become a child class such as `AccountIdentitiesPetnamesBrdige` and we'd register the listeners using the `init` method provided by `AbstractPetnamesBridge`.

matthewwalsh0 commented on app/scripts/lib/setupAccountLabelsPetnamesBridge.ts:
```diff
@@ -0,0 +1,46 @@
+  const bridge = new OneWayNameBridge(nameController, () =>
+    selectAccountLabelEntries(preferencesController.store.getState()),
+  );
+  preferencesController.store.subscribe(bridge.synchronize);
```
In the abstract class approach, this preferences controller subscription would be encapsulated inside the `abstract onSourceChange` implementation for this child class.

matthewwalsh0 commented on app/scripts/controllers/preferences.d.ts:
Love this, a great idea to define the missing types we need but in a reusable way.

Concerning the file extension, I think we reserve `.d.ts` for type declaration files for third party packages in the `types` directory, so this could just be `preferences-types.ts`?

dbrans replied:
Using .d.ts files alongside .js files is a strategy for easing a JavaScript to TypeScript migration. For example, when preferences.js is converted to preferences.ts, we can simply eliminate preferences.d.ts. This ensures that types imported in other files remain functional without needing to change import statements – they'll already be correctly written as import {Foo} from 'preferences'.

However, I get that if this approach isn't our standard practice, it might be best to discuss its potential benefits with the core team before including it in a PR.

matthewwalsh0 replied:
I didn't spot it was shadowing the existing controller file, very nice!

If this works with the current build without issue, I really like it ❤️

dbrans committed:
Merge remote-tracking branch 'origin/develop' into dbrans/petnames-accountids

dbrans committed:
snapshot

dbrans committed:
snapshot

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+  DELETED = 'DELETED',
+}
+
+type SyncDirection = 'Source->Petnames' | 'Petnames->Source';
```
As with the `ChangeType`, is it more idiomatic to use enums for fixed strings rather unions?

dbrans replied:
Done in next update.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+
+  #nameController: NameController;
+
+  #synchronizingDirection: SyncDirection | 'NONE' = 'NONE';
```
Rather than having a `NONE` entry, is it simpler to set it to `null` or `undefined` since the platform provides that state for us?

dbrans replied:
Done in next update.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+  protected abstract getSourceEntries(): PetnameEntry[];
+
+  /**
+   * Update the Source with the given entry. To be overridden by subclasses.
```
```suggestion
   * Update the Source with the given entry. To be overridden by two-way subclasses.
```

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+  #synchronize(direction: SyncDirection): void {
+    if (this.#synchronizingDirection === direction) {
+      throw new Error(
+        `synchronize(${direction}) called recursively in same direction`,
```
```suggestion
        `Attempted to synchronize recursively in same direction: ${direction}`,
```

dbrans replied:
Done in next update

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+  #getPetnameEntries(): PetnameEntry[] {
+    const { names } = this.#nameController.state;
+    const entries: PetnameEntry[] = [];
+    for (const type of Object.values(NameType)) {
```
These bridges are all specific to `NameType.ETHEREUM_ADDRESS` so we don't need to query the names for any other name types.

Arguably we could call this `AbstractAddressPetnamesBridge` but that feels like overkill as it's the only current `NameType`.

matthewwalsh0 replied:
We don't need to support any other `NameType` for our uses cases, but if this is now supported also... ❤️

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+  #applyChangeList(changeList: ChangeList): void {
+    const applyChange =
+      this.#synchronizingDirection === 'Source->Petnames'
+        ? this.updateSourceEntry.bind(this)
```
Am I missing something or is this ternary the wrong way round?

dbrans replied:
Good catch. Done in next update.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+      }
+    });
+
+    if (!this.#isTwoWay) {
```
Do we now want the inverse here?

As we want address book removals (two way) to be removed from petnames and vice versa, but for account identities (one way) we don't want anything never in the account identities to be removed from petnames?

dbrans replied:
Yep, good catch. Done in next update.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,240 @@
+import { RestrictedControllerMessenger } from '@metamask/base-controller';
+
+// Use the same type for both the source entries and the argument to NameController::setName.
+export type PetnameEntry = SetNameRequest;
```
❤️

dbrans replied:
I made a small change here.

matthewwalsh0 commented on app/scripts/lib/AccountIdentitiesPetnamesBridge.ts:
```diff
@@ -0,0 +1,52 @@
+  protected getSourceEntries(): PetnameEntry[] {
+    const { identities } = this.#preferencesController.store.getState();
+    return Object.values(identities).map((identity: AccountIdentityEntry) => ({
+      value: identity.address,
```
Do we need to normalize these addresses?

Are they always stored in lowercase or the checksummed address like the address book?

dbrans replied:
I checked it out, and here's what I found: address normalization is already taken care of by the name controller when setting a name [1], and by the useName hook when reading a name [2]. Looks like we're all set on this front, so no extra steps needed here.

[1] https://github.com/MetaMask/core/blob/be55cebdc84cb7d7f897ce7f8bfe0279bdecfdc1/packages/name-controller/src/NameController.ts#L383-L384
[2] https://github.com/MetaMask/metamask-extension/blob/8b89c3e7dbfd9d5b7317e73dc84ff2000f5b7fac/ui/hooks/useName.ts#L8-L9

matthewwalsh0 replied:
Sorry for the confusion, I'm referring to the case of the account identity addresses.

We compare them in `#computeChangeList` using their keys, so do we want to ensure the source entries are lowercase since the petnames addresses will be lowercase?

matthewwalsh0 replied:
Just highlighting this comment, thanks.

dbrans replied:
Great catch, Matt. Given that this detail can be overlooked (as I just did), would it make sense to handle the lower-casing in the AbstractPetnameBridge, (in "getKey", for example)?

matthewwalsh0 replied:
Absolutely, good idea!

dbrans replied:
Done.

dbrans replied:
Done.

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -1794,6 +1795,11 @@ export default class MetamaskController extends EventEmitter {
       }),
     }).init();
+
+    setupAccountLabelsPetnamesBridge(
```
Apologies if this was not yet refactored, but we need to update this to create an instance of the new `AccountIdentitiesPetnamesBrdige` and invoke `init`.

matthewwalsh0 commented on ui/hooks/useName.ts:
```diff
@@ -23,6 +23,8 @@ function getVariationKey(type: string, chainId: string): string {
   }
 }
 
+export const FALLBACK_VARIATION = '*';
```
We can't reference this in our background classes as we want the UI and background to be decoupled.

We could define it in the `shared` directory but since we know this will ultimately be defined in the `NameController`, we could also just define it here and in the bridge.

dbrans replied:
Done.

matthewwalsh0 commented on ui/hooks/useName.ts:
```diff
@@ -33,7 +35,10 @@ export function useName(
   const variationKey = variation ?? typeVariationKey;
-  const nameEntry = names[type]?.[normalizedValue]?.[variationKey];
+  const variationsToNameEntries = names[type]?.[normalizedValue] || {};
+  const nameEntry =
```
The huge upside to this approach is we only have to define each account identity petname once in the state, but the downside is these can't currently be deleted from the petnames UI since that is currently using the selected chain as the variation when saving.

I don't think it's worth adding more to this ticket, but something worth discussing for a potential additional ticket or future work.

dbrans replied:
Acknowledged. Created https://github.com/orgs/MetaMask/projects/47/views/1?pane=issue&itemId=48025425 for further discussion.

dbrans committed:
snapshot

dbrans committed:
snapshot

dbrans committed:
snapshot

dbrans committed:
petnames bridge refactor

dbrans committed:
Merge remote-tracking branch 'origin/develop' into dbrans/petnames-accountids

dbrans committed:
Fixed migration

dbrans committed:
snapshot

dbrans committed:
Merge remote-tracking branch 'origin/develop' into dbrans/petnames-accountids

dbrans committed:
snapshot

dbrans committed:
remove .vscode/launch.json from repo

dbrans committed:
remove .vscode/launch.json from repo

dbrans committed:
fix import type in ppom-middleware

dbrans committed:
Added launch.json configuration to debug Jest

dbrans committed:
Added launch.json configuration to debug Jest

dbrans committed:
fix lint

dbrans committed:
update to name-controller v4

dbrans committed:
removing launch.json

dbrans committed:
Merge branch 'dbrans/name-controller-v4' into dbrans/petnames-accountids

dbrans committed:
Incorporate preview build of https://github.com/MetaMask/core/pull/3705

dbrans committed:
yarn dedupe

dbrans committed:
Merge branch 'dbrans/name-controller-v4' into dbrans/petnames-accountids

dbrans committed:
yarn lavamoat:auto

dbrans committed:
Merge branch 'dbrans/name-controller-v4' into dbrans/petnames-accountids

dbrans committed:
Fixing type errors related to SetNameRequest changes

dbrans commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
revert .gitignore to develop

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
remove preview build resolution

dbrans committed:
remove preview build resolution continued

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
snapshot

dbrans committed:
Use origin in AbstractPetnamesBridge and subclasses

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
fix error

dbrans committed:
fixing useName.test

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
remove mmcore preview build

dbrans committed:
update yarn.lock

dbrans committed:
update to name-controller@4.2

dbrans committed:
yarn dedupe

dbrans commented on ui/hooks/useName.ts:
```diff
@@ -33,7 +37,10 @@ export function useName(
-  const nameEntry = names[type]?.[normalizedValue]?.[variationKey];
+  const variationsToNameEntries = names[type]?.[normalizedValue] || {};
+  const nameEntry =
+    variationsToNameEntries[variationKey] ??
```
Done.

dbrans committed:
added tests for fallback variation to useName.tests.ts

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
updating tests to mock messenger.registerInitialEventPayload

dbrans committed:
yarn.lock

dbrans committed:
update BaseControllerV2 to 4.2.0

dbrans committed:
yarn dedupe

matthewwalsh0 commented on ui/hooks/useName.test.ts:
```diff
@@ -95,17 +112,75 @@ describe('useName', () => {
+      });
+    });
+
+    it('gets used if specified variation has cleared name.', () => {
```
Super minor, but our standard is no `.` at the end of test names.

dbrans replied:
Done.

matthewwalsh0 commented on app/scripts/lib/AbstractPetnamesBridge.ts:
```diff
@@ -0,0 +1,256 @@
+            name: entry.name,
+            variation,
+            sourceId: entry.sourceId ?? undefined,
+            origin: entry.origin ?? undefined,
```
Just to confirm, we're now including the origin in both the source and target entries, but we're not yet filtering the name controller entries based on the origin to support synchronising the delete of the account identities?

This could be done using a new optional `getFilteredTargetEntries` abstract method that can be overridden to filter the current name controller entires in each concrete bridge before comparison. In the case of the account identity bridge, we could just filter by origin.

But I definitely vote to do that in a subsequent PR so we can get this beautiful but big PR in first 😄

matthewwalsh0 replied:
Forgive me, but I can't find these changes?

I'm expecting some filtering logic in `AccountIdentitiesPetnamesBrdige` in a new abstract / override method?

dbrans replied:
I apologize for the confusion caused by my previous message. Please disregard it. I've moved the changes to https://github.com/MetaMask/metamask-extension/pull/22600.

dbrans committed:
addressing comments

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
name is cleared but there are still proposed names – fallback variation is not used

dbrans commented on ui/hooks/useName.ts:
```diff
@@ -33,11 +37,20 @@ export function useName(
+  const variationsToNameEntries = names[type]?.[normalizedValue] ?? {};
+  const nameEntry =
+    variationsToNameEntries[variationKey] ??
+    variationsToNameEntries[FALLBACK_VARIATION];
```
@matthewwalsh0 suppose `variationsToNameEntries[variationKey]` is:
- `name===null` 
- has one or more `proposedNames`

Should we:
1. return the nameEntry as-is? OR
2. check the fallbackEntry for a name and merge it into the returned nameEntry?

dbrans replied:
Done.

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

matthewwalsh0 commented on app/scripts/migrations/108.ts:
```diff
@@ -0,0 +1,78 @@
+    nameEntry[FALLBACK_VARIATION] = {
+      name: accountEntry.name,
+      sourceId: null,
+      proposedNames: {},
```
Are we missing the `origin` property here to record these came from account identities?

dbrans replied:
Good spot.

matthewwalsh0 commented:
Just highlighting this comment which have got lost:

https://github.com/MetaMask/metamask-extension/pull/21956#discussion_r1446447382

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

chloeYue committed:
Account list e2e test: split account details e2e tests into multiple files and improve test names (#22635)

## **Description**
This PR dose two things:

- Split the tests in previous `account-details.spec.js` file into
multiple files and placed them in one account-menu folder. Since some
tests weren't actually belong to account details being in the
account-details file, I split them into new test files for pin/unpin and
hide/unhide actions.

- I also enhanced some test names to be more descriptive.

## **Related issues**

Fixes: #22565 

## **Manual testing steps**
Run the e2e test.

## **Screenshots/Recordings**

### **Before**


### **After**

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

PrgrmrHarshShukla committed:
Part of #20163 for file ui\components\app\terms-of-use-popup\terms-of-use-popup.js (#20618)

segun committed:
FEAT: Update what's new and settings copy to communicate additional network coverage on extension (#22618)

## **Description**
- What's new copy should be updated to:
> Steer clear of known scams while still preserving your privacy with
security alerts powered by Blockaid. This feature is available on
Arbitrum, Avalanche, BNB chain, Ethereum Mainnet, Linea, Optimism and
Polygon.
> 
> Always do your own due diligence before approving requests.

- Settings copy should be updated to:
> Privacy preserving - no data is shared with third parties. Available
on Arbitrum, Avalanche, BNB chain, Ethereum Mainnet, Linea, Optimism and
Polygon.

## **Related issues**

Fixes:
[#1695](https://github.com/MetaMask/MetaMask-planning/issues/1695)

Blocked By: #22070 

## **Screenshots/Recordings**

### **Before**


https://github.com/MetaMask/metamask-extension/assets/44811/0e6b6103-efb0-4ed8-94c2-44ac0f281ff8

### **After**


https://github.com/MetaMask/metamask-extension/assets/44811/36740992-0af2-44c5-b62b-0ac522df4d17

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

digiwand committed:
fix: Blockaid What's New os theme -> dark mode (#22649)

## **Description**

Fix Blockaid What's New image when user setting is OS and their OS
setting is dark mode

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/22646
Related to: https://github.com/MetaMask/metamask-extension/pull/22613

## **Manual testing steps**

1. Have a new instance of the wallet to populate the What's New modal
2. Test the UI with different settings and when OS is light and dark
mode


## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

darkwing committed:
UX: Send Flow: Add 'Sent to Contract' warning to new send page (#22551)

## **Description**

This PR implements the "You're sending to a contract" acknowledgement in
the new send flow. This existed in the old send flow.

## **Related issues**

N/A

## **Manual testing steps**

1. Go to the new send page
2. As the recipient, paste: `0x514910771af9ca656af840dff83e8264ecf986ca`
; this is the $LINK token address
3. See the warning, note that you cannot submit the send until
acknowledgement

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

N/A

### **After**

<img width="415" alt="SCR-20240116-mthm"
src="https://github.com/MetaMask/metamask-extension/assets/46655/87955c6d-262b-4c60-82c6-71a295f42e41">


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

montelaidev committed:
feat: check for ledger error codes in the message and display a more … (#21038)

## **Description**

This pr adds more user friendly error messages instead of just error
codes when a user encounters ledger connection issues.

Resolves https://github.com/MetaMask/metamask-extension/issues/17606

## **Manual testing steps**

1. Go to Add hardware wallet
2. On the ledger device, unlock it.
3. On the `Connect a hardware wallet` page, Click on Ledger and then
continue
4. Do not click connect when the webhid connection window pops up.
5. On the ledger device, lock the device
6. On the browser, click on the nano and then continue
7. See the new error


## **Screenshots/Recordings**

### **Before**


![ledger-error-2](https://github.com/MetaMask/metamask-extension/assets/96463427/67d44ae3-6967-415a-b2ae-bfb4e20911fd)


### **After**


![ledger-error](https://github.com/MetaMask/metamask-extension/assets/96463427/896edf82-57b4-4fe1-b04a-75afcfeed149)


## **Related issues**

Resolves https://github.com/MetaMask/metamask-extension/issues/17606

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>

Gudahtt committed:
chore: Upgrade `@metamask/controller-utils` to `v8.0.1` (#22500)

## **Description**

The package `@metamask/controller-utils` has been upgraded to `v8.0.1`.
The breaking changes did not affect any of the imports used by the
extension.

See here for the changelog:
https://github.com/MetaMask/core/blob/main/packages/controller-utils/CHANGELOG.md#801

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: MetaMask Bot <metamaskbot@users.noreply.github.com>
Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>

segun committed:
feat: Enable ppom on other chains (#22633)

makoshan committed:
Add imToken Wallet support with EIP-4527 (#21804)

## **Description**
Hello, this is the imToken team. 

imToken is a mobile crypto wallet founded in 2016. It has been operating
safely and steadily for 7 years. Recently, we successfully supported
ERC-4527, and imToken can function as a QR code-based signer now. This
integration seamlessly aligns with Metamask's bidirectional QR account
feature, enhancing the ability to track accounts whose private keys are
stored on external devices.

Therefore, we propose that Metamask expand its supported QR code-based
wallet connection methods and include imToken. This will not only
enhance Metamask's functionality and user coverage, but also provide
users with a wider, more secure and efficient choice of wallets.

We have successfully completed the integration of ERC-4527 and look
forward to discussing in depth the possibility of adding imToken into
Metamask's QR code-based wallet connection methods. Thank you very much
for your attention and feedback on this matter.

## **Related issues**
no
## **Manual testing steps**
1. Please check if the added links to the imToken official website and
tutorials are correct.
2. The process is consistent with the current QR-Based Keystone and
Airgap methods. Please refer to the test video provided below for
reference.

### Video
[Dropbox
link](https://www.dropbox.com/scl/fi/kbfcthouwgqaaueepsslc/imToken_Connect_Metamask_.mp4?rlkey=vz6js9g81oevg2y77vkizfefq&dl=0)
### User Flow

![metamask](https://github.com/MetaMask/metamask-extension/assets/7024451/9280a9dd-de4e-4635-b3db-eb44f3276b6b)

## **Screenshots/Recordings**
### **Before**
<img width="571" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/7024451/e67a1c52-4c1a-46eb-8140-8df9acc8556d">

### **After**
<img width="346" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/7024451/3246af50-7187-49a4-b7f5-a84eaf1874f7">

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@gmail.com>
Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>

segun committed:
feat: Add PPOM Validation to normal send. (#22321)

## **Description**

Currently we send ppom validation requests only for incoming requests,
this PR adds ppom validation for send requests from wallets too.

## **Related issues**

Fixes:
[#1657](https://github.com/MetaMask/MetaMask-planning/issues/1657)

## **Manual testing steps**

1. Build and launch MM
2. Send ETH on mainnet to this address
`0x5FbDB2315678afecb367f032d93F642f64180aa3`
3. See that Blockaid banner is not shown
4. Checkout this branch, build and launch MM
5. Send ETH on mainnet to the address
`0x5FbDB2315678afecb367f032d93F642f64180aa3`
6. See that Blockaid banner is shown.

## **Screenshots/Recordings**

### **Before**


https://github.com/MetaMask/metamask-extension/assets/44811/9c81e157-1d78-4609-a037-7a8897309d12

### **After**


https://github.com/MetaMask/metamask-extension/assets/44811/8b2ec828-9636-4302-a943-b023caaba730

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

salimtb committed:
feat: hardcode svg network (#22643)

## **Description**

hardcode a mapping of chainIDs to currency symbol, once the svg are
uploaded we mapped the images to the network name and chainIDs

**Acceptance criteria**
When the network list is displayed within the Select a network modal, or
custom network list, or any similar context, the hardcoded network image
avatars should be displayed. This will reduce the often appearing "?"
that appears when the network avatar is not present.

## **Related issues**

Fixes:
[#1813](https://github.com/MetaMask/MetaMask-planning/issues/1813)

## **Manual testing steps**

1. Go to the network modal
2. Add a new network displayed on this
[list](https://docs.google.com/spreadsheets/d/16G3SRiQvWXK6hp5duob0aopeKyx2HOAQpz152Fy8VNQ/edit#gid=2080069111)
3. check if the image is displayed

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**


https://github.com/MetaMask/metamask-extension/assets/26223211/b109668a-9bc2-432b-b603-e90e9c85804f



### **After**



https://github.com/MetaMask/metamask-extension/assets/26223211/aae091f7-a89c-4653-a996-223ec288fd31


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

montelaidev committed:
feat: update UI elements to use `getSelectedInternalAccount` selector (#21553)

## **Description**
Currently, MetaMask faces tight coupling between its UI and the
keyring-controller. The UI heavily relies on the keyring-controller's
state while also amalgamating information from various sources, such as
preferences and balances.

However, this approach presents some challenges. The keyring-controller
must be aware of the UI's limitations, like the need for instant account
list provision to avoid lag. Moreover, it takes on the responsibility of
adding metadata to accounts, such as the associated keyring type,
required for displaying account details.

To address these issues, the introduction of the accounts-controller
comes into play as a new abstraction layer between the UI and the
keyring-controller. This separation of responsibilities allows the
keyring-controller to focus on two main tasks:

- Routing requests to the appropriate keyring.
- Persisting the state of the keyrings.

This PR replaces the usage of `getSelectedAddress` with
`getSelectedInternalAccount`

Depends on: https://github.com/MetaMask/metamask-extension/pull/21554

## **Related issues**

Resolves https://github.com/MetaMask/accounts-planning/issues/135

## **Manual testing steps**

1. Use a hardware wallet to test all the flows

## **Screenshots/Recordings**

No UI/UX changes

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: MetaMask Bot <metamaskbot@users.noreply.github.com>
Co-authored-by: gantunesr <gustavoantunesrod@gmail.com>
Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>
Co-authored-by: Howard Braham <howrad@gmail.com>
Co-authored-by: chloeYue <105063779+chloeYue@users.noreply.github.com>
Co-authored-by: Harsh Shukla <125105825+PrgrmrHarshShukla@users.noreply.github.com>
Co-authored-by: Olusegun Akintayo <akintayo.segun@gmail.com>
Co-authored-by: Ariella Vu <20778143+digiwand@users.noreply.github.com>
Co-authored-by: David Walsh <davidwalsh83@gmail.com>
Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>
Co-authored-by: Mark Stacey <markjstacey@gmail.com>
Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>
Co-authored-by: Mako Shan <imakoshan@gmail.com>
Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@gmail.com>

martykan committed:
fix: Trezor connect - update to v9 in trezor-usb-permissions (#22591)

## **Description**

We found some issues at Trezor with the Metamask integration, one of
them related to the USB permissions page being linked to a deprecated
version of our SDK.
Metamask already uses the v9 package, but in this particular instance
it's hardcoded to load from v5.

We added a workaround on our side for now, by adding a redirect. 
This PR solves the problem properly by updating the version in the code.

# **Related issues**

Unknown

## **Manual testing steps**

NA

## **Screenshots/Recordings**

NA

Co-authored-by: Brad Decker <bhdecker84@gmail.com>
Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>
Co-authored-by: Sébastien Van Eyck <sebastien.vaneyck@consensys.net>

dan437 committed:
Update Gas API URL (#22544)

chloeYue committed:
E2E test: Improve fill method in driver (#22652)

## **Description**

When investigating flaky tests, I found that some tests are flaky
because the input field was not cleared before entering a new value. The
issue is that Selenium sometimes doesn't clear input fields effectively.
Now, we use sendkeys() to clear input fields, i've added another way to
use selenium actions to simulate we select all text with keyboard and
replace it with the new value. With both ways, we can clear input fields
more effectively, which I believe will stabilize multiple flaky tests.

Examples:

Failing test because input field was not cleared as expected:
![Screenshot 2024-01-24 at 11 46
01](https://github.com/MetaMask/metamask-extension/assets/105063779/5f067bd0-c426-42fd-b378-62b6dbba6b8a)


Tests failing because although the input field was cleared, the
application automatically added a '0'. So when we entered the input
value '4000', it was actually entered as '04000'. With my approach, when
we find that the input field is not empty, it will select the '0' and
replace it with '4000'. This way, we finally get the correct value of
'4000'.
<img width="363" alt="Screenshot 2024-01-24 at 14 44 04"
src="https://github.com/MetaMask/metamask-extension/assets/105063779/98d6d811-46df-4d15-be50-7f18fe13fdf7">


I also fixed another flaky test in this PR.

## **Related issues**

Fixes: #22573 

## **Manual testing steps**

Run the e2e test.

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
normalizing (lowercase) addresses in abstract bridge

dbrans committed:
fixing useName to return fallback if variation has no name

dbrans committed:
adding origin to migration

dbrans committed:
fixing tests

dbrans committed:
fix lint

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
readability

dbrans committed:
Merge branch 'develop' into dbrans/petnames-accountids

dbrans committed:
reverting app/scripts/lib/keyring-snaps-permissions.test.ts
